### ISSUE TITLE
readinessProbe required in kube-apiserver

### ISSUE THREAD

**Issue by bergerx** [gardener/gardener#230]
So this only an idea for improving the Shoot creation flow, an this would be a nice to have feature to troubleshoot provisioning issues and and early break before creating resources that could not be used.

Multiple time we had shoot's kube-apiserver failing to reach to the etcd (still investigating the root cause).

Currently [charts/seed-controlplane/charts/kube-apiserver/templates/kube-apiserver.yaml](../blob/release-0.7/charts/seed-controlplane/charts/kube-apiserver/templates/kube-apiserver.yaml#L92) doesn't have a `readinessProbe`.

And the [WaitUntilKubeAPIServerReady](https://github.com/gardener/gardener/blob/release-0.7/pkg/operation/botanist/waiter.go#L49) step is not really meaningful without a readiness probe since if a Container does not provide one, the default state is Success. So that is not really checking for Readiness but only existence of the container.

**Comment by rfranzke** [comment]
Which kind of `readinessProbe` do you suggest for the `kube-apiserver` deployment? It does not expose a `/ready` endpoint but only a `/healthz` route. `/healthz/etcd` could be an option?!
But what is the expected outcome of that? When it reports that it is not ready, then it won't get any traffic. Hence, in the end the behaviour is the same compared to now - no operation involving it will work.

**Comment by bergerx** [comment]
If the current `/healthz` endpoint runs below checks, isnt it already quite suitable for readiness as is:

```
[+]ping ok
[-]etcd failed: reason withheld
[+]poststarthook/generic-apiserver-start-informers ok
[+]poststarthook/start-apiextensions-informers ok
[+]poststarthook/start-apiextensions-controllers ok
[+]poststarthook/bootstrap-controller ok
[+]poststarthook/rbac/bootstrap-roles ok
[+]poststarthook/ca-registration ok
[+]poststarthook/start-kube-apiserver-informers ok
[+]poststarthook/start-kube-aggregator-informers ok
[+]poststarthook/apiservice-registration-controller ok
[+]poststarthook/apiservice-status-available-controller ok
[+]poststarthook/apiservice-openapi-controller ok
[+]poststarthook/kube-apiserver-autoregistration ok
[+]autoregister-completion ok
```

**Comment by rfranzke** [comment]
Sure, but I've not understood what you expect to happen. When the readiness probe in Kubernetes fails then no traffic will be routed to the underlying pods. Hence, the outcome of the operation Gardener wants to perform is the same as it is now - it won't work / it won't get better. Or do you only suggest to use it for improving the step in which the operation fails (`WaitUntilKubeAPIServerIsReady` instead of a following)? That we could definitely do, agreed!

**Comment by swapnilgm** [comment]
Although there isn't any direct api on apiserver(say /ready), and `/healthz` is only health endpoint and doesn't mean apiserver is completely ready to accept and process request,  I think having readiness probe using `/healthz` makes sense. This will halt the flow at `WaitUntilKubeAPIServerIsReady` step. And help in early detection of error like etcd issues.

Said that, this surely doesn't solve the problem completely though, cause there are always chances on apiserver passing health check i.e new readiness probe at `WaitUntilKubeAPIServerIsReady` but failing it before next step like `reconcilemachines` which will end up not registering root cause in shoot status for that iteration of reconciliation.

**Comment by bergerx** [comment]
From @rfranzke:
> Sure, but I've not understood what you expect to happen. When the readiness probe in Kubernetes fails then no traffic will be routed to the underlying pods.

Since there is only a liveness probe and no readiness probe, as soon as the container is started, it immediately transitions into Ready state, after some time Liveness probe will fail and restart the kube-apiserver, until then kube-apiserver will be assumed ready to serve. `WaitUntilKubeAPIServerIsReady` checks for Ready state not Liveness. Having a readiness probe will block flow at `WaitUntilKubeAPIServerIsReady`, and we won't see indirect errors reported at later steps. e.g currently in one of our clusters we never had a *really* `Ready` kube-apiserver, it kept being restarted as result of liveness prove, but flow is not blocked at `WaitUntilKubeAPIServerIsReady` step, and eventually blocked on `ReconcileMachines` step as there is no Healthy apiserver's to reach.

**Comment by rfranzke** [comment]
OK, thank you for the elaboration. Will incorporate it! Great feedback @bergerx!
